Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/CacheHandler.kt

```diff
@@ -0,0 +1,802 @@
+                            "Couldn't create cache file meta, path = ${cacheFileMeta.getFullPath()}")
+                }
+
+                val result = updateCacheFileMeta(
+                        createdFile,
+                        true,
+                        System.currentTimeMillis(),
+                        false
+                )
+
+                if (!result) {
+                    throw IOException("Cache file meta update failed")
+                }
+            }
+
+            cacheFile
+        } catch (error: IOException) {
+            Logger.e(TAG, "Unknown error", error)
+            deleteCacheFile(cacheFile)
+            null
+        }
+    }
+
+    /**
+     * Checks whether this file is already downloaded by reading it's meta info. If a file has no
+     * meta info or it cannot be read - deletes the file so it can be re-downloaded again with all
+     * necessary information
+     * */
+    fun isAlreadyDownloaded(file: RawFile): Boolean {
+        return try {
+            if (!fileManager.exists(file)) {
+                return false
+            }
+
+            if (!fileManager.getName(file).endsWith(CACHE_EXTENSION)) {
+                Logger.e(TAG, "Not a cache file! file = " + file.getFullPath())
+                return false
+            }
+
+            val cacheFileMetaFile = getCacheFileMetaByCacheFile(file)
+            if (cacheFileMetaFile == null) {
+                Logger.e(
+                        TAG, "Couldn't get cache file meta by cache file, " +
+                        "file = ${file.getFullPath()}"
+                )
+                return false
+            }
+
+            if (!fileManager.exists(cacheFileMetaFile)) {
+                Logger.e(
+                        TAG, "Cache file meta does not exist, " +
+                        "cacheFileMetaFile = ${cacheFileMetaFile.getFullPath()}"
+                )
+
+                deleteCacheFile(file)
+                return false
+            }
+
+            if (fileManager.getLength(cacheFileMetaFile) <= 0) {
+                // File is empty
+                deleteCacheFile(file)
+                return false
+            }
+
+            val cacheFileMeta = readCacheFileMeta(cacheFileMetaFile)
+            if (cacheFileMeta == null) {
+                deleteCacheFile(file)
+                return false
+            }
+
+            cacheFileMeta.isDownloaded
+        } catch (error: Throwable) {
+            Logger.e(TAG, "Unknown error", error)
+            deleteCacheFile(file)
+            false
+        }
+    }
+
+    /**
+     * Marks the file as a downloaded (sets a flag in it's meta info). If meta info cannot be read
+     * deletes the file so it can be re-downloaded again.
+     * */
+    fun markFileDownloaded(output: RawFile): Boolean {
+        return try {
+            if (!fileManager.exists(output)) {
+                Logger.e(TAG, "File does not exist! file = ${output.getFullPath()}")
```

Comment:
Should probably also add deleteCacheFile() here to delete the cache file meta in case it still exists.